File: app/src/main/java/com/memtrip/pinyin/audio/stream/StreamingNavigator.java
package com.memtrip.pinyin.audio.stream;

import android.content.Context;
import android.content.Intent;

import static com.memtrip.pinyin.audio.stream.StreamingAction.ACTION_NOTIFICATION_PLAY;

public abstract class StreamingNavigator<T extends Stream> implements Navigator<T> {

    private final StreamIntent<T> streamIntent;
    private final Class<? extends StreamingService> classDef;

    public StreamingNavigator(StreamIntent<T> streamIntent,
                              Class<? extends StreamingService> classDef) {

        this.streamIntent = streamIntent;
        this.classDef = classDef;
    }

    @Override
    public void play(T data, Context context) {
        Intent intent = new Intent(context, classDef);
        intent.setAction(ACTION_NOTIFICATION_PLAY.toString());

        streamIntent.into(data, intent);

        context.startService(intent);
    }
}

File: app/src/main/java/com/consistence/pinyin/audio/stream/StreamingNavigator.kt
package com.consistence.pinyin.audio.stream

import android.content.Context
import android.content.Intent

abstract class StreamingNavigator<T : Stream>(private val streamIntent: StreamIntent<T>,
                                              private val classDef: Class<out StreamingService<T>>) : Navigator<T> {

    override fun play(data: T, context: Context) {
        val intent = Intent(context, classDef)
        intent.action = ACTION_NOTIFICATION_PLAY

        streamIntent.into(data, intent)

        context.startService(intent)
    }

    companion object {
        private const val ACTION_NOTIFICATION_PLAY = "ACTION_NOTIFICATION_PLAY"
    }
}